src/t_hash.c

```diff
@@ -492,6 +418,111 @@ int hashTypeSet(robj *o, sds field, sds value, int flags) {
+ * returns 1 if the expiration time was applied.
+ * returns 2 when 'expire' indicate a past Unix time. In this case, if the item exists in the HASH, it will also be expired.
+ */
+int hashTypeSetExpire(robj *o, sds field, long long expiry, int flag) {
+    /* If no object we will return -2 */
+    if (o == NULL) return -2;
+
+    if (o->encoding == OBJ_ENCODING_LISTPACK) {
+        unsigned char *vstr;
+        unsigned int vlen;
+        long long vll;
+        /* We do not want to convert to listpack for no good reason.
+         * So we first check if the item exists.*/
+        if (hashTypeGetFromListpack(o, field, &vstr, &vlen, &vll) < 0) {
+            return -2;
+        }
+        /* When listpack representation is used, we consider it as infinite TTL,
+         * so expire command with gt always fail the GT as well as existence(XX).
+         * Else, we already know we are going to set an expiration so we expend to hashtable encoding. */
+        if (flag & EXPIRE_XX || flag & EXPIRE_GT) {
+            return 0;
+        } else {
+            hashTypeConvert(o, OBJ_ENCODING_HASHTABLE);
+        }
+    }
+
+    /* we must be hashtable encoded */
+    serverAssert(o->encoding == OBJ_ENCODING_HASHTABLE);
+
+    hashtable *ht = o->ptr;
+    void **entry_ref = NULL;
+    if ((entry_ref = hashtableFindRef(ht, field))) {
+        entry *current_entry = *entry_ref;
+        long long current_expire = entryGetExpiry(current_entry);
+        if (flag) {
+            /* NX option is set, check no current expiry */
+            if (flag & EXPIRE_NX) {
+                if (current_expire != EXPIRY_NONE) {
+                    return 0;
+                }
+            }
+
+            /* XX option is set, check current expiry */
+            if (flag & EXPIRE_XX) {
+                if (current_expire == EXPIRY_NONE) {
+                    return 0;
+                }
+            }
+
+            /* GT option is set, check current expiry */
+            if (flag & EXPIRE_GT) {
+                /* When current_expire is -1, we consider it as infinite TTL,
+                 * so expire command with gt always fail the GT. */
+                if (expiry <= current_expire || current_expire == EXPIRY_NONE) {
+                    return 0;
+                }
+            }
+
+            /* LT option is set, check current expiry */
+            if (flag & EXPIRE_LT) {
+                /* When current_expire -1, we consider it as infinite TTL,
+                 * so if there is an expiry on the key and it's not less than current, we fail the LT. */
+                if (current_expire != EXPIRY_NONE && expiry >= current_expire) {
+                    return 0;
+                }
+            }
+        }
+        *entry_ref = entrySetExpiry(current_entry, expiry);
+        hashTypeTrackUpdateEntry(o, current_entry, *entry_ref, current_expire, expiry);
+        return 1;
+    }
+    return -2; // we did not find anything to do. return -2
```

Comment:
please use symbolic names for these values of -2, -1, 0, 1

Reply:
done